WordPress/src/test/java/org/wordpress/android/ui/qrcodeauth/QRCodeAuthViewModelTest.kt

```diff
@@ -1,30 +1,553 @@
+        job2.cancel()
+    }
+
+    @Test
+    fun `given valid qr code, when scanned qrcode, then validated is shown`() = runBlockingTest {
+        initValidate()
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(VALIDATED)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given invalid host, when scanned qrcode, then error is shown`() = runBlockingTest {
+        whenever(validator.isValidUri(scannedValue)).thenReturn(false)
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(INVALID_DATA)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given invalid query params, when scanned qrcode, then error is shown`() = runBlockingTest {
+        whenever(validator.isValidUri(scannedValue)).thenReturn(true)
+        whenever(validator.extractQueryParams(scannedValue)).thenReturn(invalidQueryParams)
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(INVALID_DATA)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given not authorized error, when validate failure, then auth failed is shown`() = runBlockingTest {
+        initValidate(false, NOT_AUTHORIZED)
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(AUTH_FAILED)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given error, when validate failure, then invalid data is shown`() = runBlockingTest {
+        initValidate(false, GENERIC_ERROR)
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(INVALID_DATA)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given data invalid, when validate failure, then expired is shown`() = runBlockingTest {
+        initValidate(false, DATA_INVALID)
+
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+        viewModel.onScanSuccess(scannedValue)
+
+        assertThat(result.last().type).isEqualTo(EXPIRED)
+
+        job.cancel()
+    }
```

Comment:
Based on `Test B: Scanning a valid but expired token` result [here](https://github.com/wordpress-mobile/WordPress-Android/pull/16713#issuecomment-1150963922), this test case might need to be updated to replace `DATA_INVALID` with `AUTHORIZATION_REQUIRED`.

Reply:
@ashiagr - see note below about AUTHORIZATION_REQUIRED. This test should be fine as written. Thanks